x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/loadingservice/ModelLoadingService.java

```diff
@@ -377,34 +402,41 @@ private void handleLoadSuccess(String modelId,
             trainedModelConfig.getLicenseLevel(),
             modelStatsService,
             trainedModelCircuitBreaker);
-        boolean modelAcquired = false;
+        final ModelAndConsumerLoader modelAndConsumerLoader = new ModelAndConsumerLoader(new ModelAndConsumer(loadedModel, consumer));
         synchronized (loadingListeners) {
-            listeners = loadingListeners.remove(modelId);
-            // if there are no listeners, simply release and leave
-            if (listeners == null) {
-                loadedModel.release();
-                return;
-            }
-
+            populateNewModelAlias(modelId);
             // If the model is referenced, that means it is currently in a pipeline somewhere
             // Also, if the consume is a search consumer, we should always cache it
-            if (referencedModels.contains(modelId) || consumer.equals(Consumer.SEARCH)) {
-                // temporarily increase the reference count before adding to
-                // the cache in case the model is evicted before the listeners
-                // are called in which case acquire() would throw.
-                loadedModel.acquire();
-                localModelCache.put(modelId, new ModelAndConsumer(loadedModel, consumer));
+            if (referencedModels.contains(modelId)
+                || Sets.haveNonEmptyIntersection(modelIdToModelAliases.getOrDefault(modelId, new HashSet<>()), referencedModels)
+                || consumer.equals(Consumer.SEARCH)) {
+                try {
+                    // The local model may already be in cache. If it is, we don't bother adding it to cache.
+                    // If it isn't, we flip an `isLoaded` flag, and increment the model counter to make sure if it is evicted
+                    // between now and when the listeners access it, the circuit breaker reflects actual usage.
+                    localModelCache.computeIfAbsent(modelId, modelAndConsumerLoader);
+                } catch (ExecutionException ee) {
+                    logger.warn(() -> new ParameterizedMessage("[{}] threw when attempting add to cache", modelId), ee);
+                }
                 shouldNotAudit.remove(modelId);
```

Comment:
@davidkyle This was a bit tricky. It is possible that we don't have any listeners, but still load something into cache. This would usually be due to loading a new model via a name change and NOT a new reference and during that time, no callers are requesting that model.

Also, I noticed in testing that we were unnecessarily evicting models that were accidentally cached twice (the second cache evicts the first model cached). This causes weird logging and is ultimately unnecessary if we use the `computeIfAbsent` logic in the model cache.

I ran this a bunch locally and it is all checking out ok. CI and should time agree.

Reply:
ModelLoadingService has good test coverage CI will find any problems